src/apint/arithmetic.rs

```diff
@@ -919,68 +1791,374 @@ mod tests {
+                            assert_eq!(lhs.$fun_into(&rhs).unwrap(), ApInt::from($r3));
+                        }
+                    }
+                };
+                ($signed:expr,$fun:ident,$r0:expr,$r1:expr,$r2:expr,$r3:expr/*,$div_op:ident*/) => {
+                    {
+                        let mut lhs = ApInt::from(80i8);
+                        let mut rhs = ApInt::from(7i8);
+                        ApInt::$fun(&mut lhs, &mut rhs).unwrap();
+                        assert_eq!(lhs, ApInt::from($r0.0));
+                        assert_eq!(rhs, ApInt::from($r0.1));
+                    }
+                    if $signed {
+                        {
+                            let mut lhs = ApInt::from(80i8);
+                            let mut rhs = ApInt::from(-7i8);
+                            ApInt::$fun(&mut lhs, &mut rhs).unwrap();
+                            assert_eq!(lhs, ApInt::from($r1.0));
+                            assert_eq!(rhs, ApInt::from($r1.1));
+                        }
+                        {
+                            let mut lhs = ApInt::from(-80i8);
+                            let mut rhs = ApInt::from(7i8);
+                            ApInt::$fun(&mut lhs, &mut rhs).unwrap();
+                            assert_eq!(lhs, ApInt::from($r2.0));
+                            assert_eq!(rhs, ApInt::from($r2.1));
+                        }
+                        {
+                            let mut lhs = ApInt::from(-80i8);
+                            let mut rhs = ApInt::from(-7i8);
+                            ApInt::$fun(&mut lhs, &mut rhs).unwrap();
+                            assert_eq!(lhs, ApInt::from($r3.0));
+                            assert_eq!(rhs, ApInt::from($r3.1));
+                        }
+                    }
+                }
+            }
+            s!(false,wrapping_udiv_assign,into_wrapping_udiv,11i8,0,0,0);
+            s!(false,wrapping_urem_assign,into_wrapping_urem,3i8,0,0,0);
+            s!(true,wrapping_sdiv_assign,into_wrapping_sdiv,11i8,-11i8,-11i8,11i8);
+            s!(true,wrapping_srem_assign,into_wrapping_srem,3i8,3i8,-3i8,-3i8);
+            s!(false,wrapping_udivrem_assign,(11i8,3i8),(0,0),(0,0),(0,0));
+            s!(false,wrapping_uremdiv_assign,(3i8,11i8),(0,0),(0,0),(0,0));
+            s!(true,wrapping_sdivrem_assign,(11i8,3i8),(-11i8,3i8),(-11i8,-3i8),(11i8,-3i8));
+            s!(true,wrapping_sremdiv_assign,(3i8,11i8),(3i8,-11i8),(-3i8,-11i8),(-3i8,11i8));
         }
 
+        //NOTE: this test only works if multiplication and a few other functions work
         #[test]
-        fn with_neg() {
-            let lhs = ApInt::from(72_i32);
-            let rhs = ApInt::from(-12_i32);
-            let result = lhs.into_wrapping_sdiv(&rhs).unwrap();
-            assert_eq!(result, ApInt::from(-6_i32));
+        fn complex() {
+            //there are many special case and size optimization paths,
```

Comment:
I am always unsure as to how to structure tests for special cases. I like to put every special case in its own test so I have a better overview about what actually is broken whenever I run the tests. However, I do not want to say that this is the best procedure about test structuring and also results in many many test cases.

Reply:
There are thousands of special cases I am testing there. Some of this is actually brute force fuzzing that I designed specifically for `ApInt`s and the functions on them. I didn't encounter any really crazy edge or corner cases, but these tests are what give me a lot of confidence in my functions